File: frontend/src/types/tiles.ts
import { RangeEnum } from './range';
import { Timeframe } from './timeframe';

export enum TileType {
  Chart = 'Chart',
  TimeAndSales = 'TimeAndSales',
  AccountInfo = 'AccountInfo',
  TickerInfo = 'TickerInfo',
  Headlines = 'Headlines',
  Scanner = 'Scanner',
}

export interface TileInterface {
  id: string;
  x: number;
  y: number;
  w: number;
  h: number;
  data: TileData;
}

export type TileData =
  | {
      type: TileType.Chart;
      config: ChartConfig;
    }
  | {
      type: TileType.TimeAndSales;
      config: TimeAndSalesConfig;
    }
  | {
      type: TileType.AccountInfo;
    };

export type Configs = ChartConfig | TimeAndSalesConfig;

export interface ChartConfig {
  ticker: string;
  timeframe: Timeframe;
  range: RangeEnum;
}

export interface TimeAndSalesConfig {
  ticker: string;
}

File: pkg/models/tiles.go
package models

type Tile struct {
	ID   string   `json:"id"`
	X    int      `json:"x"`
	Y    int      `json:"y"`
	W    int      `json:"w"`
	H    int      `json:"h"`
	Data TileData `json:"data"`
}

type TileData struct {
	Type   TileType    `json:"type"`
	Config interface{} `json:"config"`
}

type TileType string

const (
	Chart           TileType = "Chart"
	TimeAndSales    TileType = "TimeAndSales"
	AccountOverview TileType = "AccountOverview"
	TickerInfo      TileType = "TickerInfo"
	Headlines       TileType = "Headlines"
	Scanner         TileType = "Scanner"
)
